src/java.base/share/native/libjava/io_util.c

```diff
@@ -92,27 +95,42 @@ readBytes(JNIEnv *env, jobject this, jbyteArray bytes,
     if (len == 0) {
         return 0;
     } else if (len > BUF_SIZE) {
-        buf = malloc(len);
+        buf_size = len < MAX_MALLOC_SIZE ? len : MAX_MALLOC_SIZE;
+        buf = malloc(buf_size);
         if (buf == NULL) {
             JNU_ThrowOutOfMemoryError(env, NULL);
             return 0;
         }
     } else {
         buf = stackBuf;
+        buf_size = BUF_SIZE;
     }
 
-    fd = getFD(env, this, fid);
-    if (fd == -1) {
-        JNU_ThrowIOException(env, "Stream Closed");
-        nread = -1;
-    } else {
-        nread = IO_Read(fd, buf, len);
-        if (nread > 0) {
-            (*env)->SetByteArrayRegion(env, bytes, off, nread, (jbyte *)buf);
-        } else if (nread == -1) {
-            JNU_ThrowIOExceptionWithLastError(env, "Read error");
-        } else { /* EOF */
+    nread = 0;
+    while (nread < len) {
+        remaining = len - nread;
+        read_size = remaining <= buf_size ? remaining : buf_size;
+        fd = getFD(env, this, fid);
+        if (fd == -1) {
+            JNU_ThrowIOException(env, "Stream Closed");
             nread = -1;
+            break;
+        }
+        n = IO_Read(fd, buf, read_size);
+        if (n > 0) {
+            (*env)->SetByteArrayRegion(env, bytes, off, n, (jbyte*)buf);
+            nread += n;
+            // Exit loop on short read
+            if (n < read_size)
+                break;
+            off += n;
+        } else if (n == -1) {
+            JNU_ThrowIOExceptionWithLastError(env, "Read error");
+            break;
+        } else { // EOF
+            if (nread == 0)
+                nread = -1;
+            break;
```

Comment:
Can you prototype doing the loop in Java rather than in native code so that there is less native code to maintain?

Reply:
I prototyped doing the read loop in Java and the performance seemed to be about the same. The problem is that the loop needs to exit when the native `read()` function performs a short read, i.e., fewer bytes are read than were requested. Without this, at least one regression test fails. The reason is not completely clear.

To detect such a short read in the Java layer would require some way for the native layer to notify the Java layer. One potential such method is
```
boolean readBytes(byte[] b, int off, int len, int[] nread) {}
```
where the return value indicates whether all or only some of the `len` bytes were read. If not all were read, then `nread[0]` would contain the number actually read or `-1`.

Another possibility is
```
int readBytes(byte[] b, int off, int len, int maxBufferSize) {}
```
which is identical to the current `readBytes()` except that the maximum transfer buffer size is specified as a method parameter instead of being defined by a symbolic constant in the native layer. In this case a short read would be detected if `len >= maxBufferSize` and the returned value is less than `maxBufferSize`.

As for the read loop being in native code, note that the write loop is also already in native code, so if the read loop were moved to Java, then probably the write loop should be as well.

One advantage of the loops being in native code is that there is only one `malloc()` per Java `read(byte[],int,int)` or `write(byte[],int,int)` invocation.